### ISSUE TITLE
Batch online phase

### ISSUE THREAD

**Issue by amiller** [initc3/HoneyBadgerMPC#22]
Right now the passive MPC online phase only handles one share at a time. As a result it has quadratic communication/computation cost. We can improve the performance through the use of batch public reconstruction, even just for the passive case. 

A batch reconstruction subprotocol is written here: https://github.com/changxu2/HoneyBadgerMPC/blob/pubrec/honeybadgermpc/batch_reconstruction.py
But now we want to incorporate it into an MPC execution context. This issue is to create a new file `passive-batch.py` which implements an MPC context that relies on batch reconstruction.

- [x] Write a ShareArray object. This can be instantiated with single Share objects.
- [x] Define an 'open()' method for a ShareArray, which breaks the array into chunks of at least f+1 and calls batch_reconstruction accordingly
- [x] We can eventually overload its operations like +/- and scalar mult, and it can be backed by a rust module with a packed encoding.
- [x] Benchmarking. Can we show how many bytes we save as a function of N by using this method?

If this works out, we can just refactor this into passive.py (so passive.py supports both single and batch).

**Comment by amiller** [comment]
Note: this issue is just to create the passive secure case, but the batch_reconstruction method already handles the malicious case (it performs B-W decoding if the first attempt fails). Maybe we go straight to active then.

**Comment by amiller** [comment]
Here's an example of a program making use of the batch interface
```python
async def openEveryPair(context, As, Bs):
   assert type(As) is type(Bs) is list and len(As) == len(Bs)
   ABs = context.ShareArray( [ A - B for A,B in zip(As,Bs) ] ) 
   return await ABs.open() # calls batch_reconstruction
```

**Comment by amiller** [comment]
Here's an example interface based on a ShareArray class.
https://github.com/amiller/HoneyBadgerMPC/blob/batch-online/honeybadgermpc/passive.py
The main next steps are:
- match the ShareArray.open method use batch reconstruction,
- make more examples (like the jubjub example) work with this batch interface

**Comment by amiller** [comment]
Benchmarking is addressed in 
https://github.com/initc3/HoneyBadgerMPC/commit/4a3cd0de39fbb860296dcbf7d8c28a00008fdef3#diff-cf737a48d52dab0d5bab7b59f3a5cf2dR7
Batch-online is merged prior to asyncmix benchmarks
https://github.com/initc3/HoneyBadgerMPC/commit/15642a51640a1fee7d66f95202567208911abe0e
We can close this once merging all these to dev tho

**Comment by amiller** [comment]
After some initial benchmarking
    https://github.com/amiller/HoneyBadgerMPC/blob/batch_bm/tests/open_secrets_bm.py
```
Test n = 16, t = 5, field = 52435875175126190479447740508185965837690552500527637822603658699938581184513
         6385472 function calls (6348556 primitive calls) in 3.365 seconds

   Ordered by: cumulative time, internal time
   List reduced from 57 to 29 due to restriction <0.5>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.006    0.006    3.365    3.365 batch_reconstruction.py:73(attempt_reconstruct_batch)
     1667    0.036    0.000    3.355    0.002 robust_reconstruction.py:15(attempt_reconstruct)
     1667    0.008    0.000    2.743    0.002 wb_interpolate.py:129(decode)
     1667    0.081    0.000    2.702    0.002 polynomial.py:62(interpolate)
```
it seems clear that our current polynomial interpolation is quite slow, and can easily become a bottleneck for relatively small sizes as `n` grows.

We should look into using an interpolation routine from NTL or Flint to accelerate it:
`ZZ_pX interpolate(const vec_ZZ_p& a, const vec_ZZ_p& b);`

https://github.com/u-u-h/NTL/blob/d71a0c893f9080648cb8fe6798f1546e1a388535/src/ZZ_pX1.c#L837

This likely blocks on #117 which will get us an interface from python to NTL.

**Comment by amiller** [comment]
We can consider this closed with #160 #152
